### ISSUE TITLE
[Bug]: updating client on Windows requires reboot no matter what.

### ISSUE THREAD

**Issue by razorbackgeek** [nextcloud/desktop#5369]
### ⚠️ Before submitting, please verify the following: ⚠️

- [X] This is a **bug**, not a question or a configuration issue.
- [X] This issue is **not** already reported on Github (I've searched it).
- [X] Nextcloud Server and Desktop Client are **up to date**. See [Server Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) and [Desktop Releases](https://nextcloud.com/install/#install-clients) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/)

### Bug description

When updating the existing Nextcloud client on Windows 10 PC it will prompt for the update.  During update, it will ask to close explorer, or you must restart PC.  I let it close explorer automatically.  Install finishes, requires reboot anyway.  After several attempts at getting around this, including manually closing explorer before upgrading it still requires a reboot.

### Steps to reproduce

1.Use the setup wizard to upgrade to current release.
2.Let it automatically close explorer, or you can manually close explorer.
3.A reboot is still required even though you took the steps to avoid reboot.

### Expected behavior

The expected behavior is to not have to reboot after installing the update.  This has happened on multiple updates all have required a reboot regardless of outcome.

### Which files are affected by this bug

Client install package

### Operating system

Windows

### Which version of the operating system you are running.

Windows 10

### Package

Appimage

### Nextcloud Server version

23.0.1

### Nextcloud Desktop Client version

3.6.6

### Is this bug present after an update or on a fresh install?

Updated from a minor version (ex. 3.4.2 to 3.4.4)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### Are you using an external user-backend?

- [X] Default internal user-backend
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

**Comment by Aldaris1985** [comment]
Confirmed also from my side. This is the behavior for a couple of versions already.

**Comment by farhills** [comment]
Confirmed on my systems (windows 10 22H2) for the last few updates as well, 3.7.3 being the latest.

**Comment by tstreibl** [comment]
same on Windows 11 during installation of 3.7.4

**Comment by r2evans** [comment]
This is a running problem, including https://github.com/nextcloud/desktop/issues/4377 and others. I just confirmed it on 3.8. The problem was allegedly fixed in 3.5.4, though by the description I'm not certain if it was supposed to not require a reboot on 3.5.4 or _after_ 3.5.4, but since I've seen "you must restart your system" every time I've upgraded, it seems to be a misunderstanding of the problem.

Please, can one of the devs (NC or volunteer) acknowledge the issue?

My concerns are at various levels of the process:

- One of the original (egregious) problems _has been resolved_, and that is restarting the computer without user notification or ability to pause/interrupt the reboot process.
- The update reports that `Explorer.exe` (and perhaps others) must be closed in order to do the upgrade, and once killed it does not ensure that explorer is started again, effectively hobbling the system. Often other apps are listed despite that no integrations with them are installed.
    Most of the time I'm able to bring back `Explorer` manually by bringing up `taskmgr.exe` and starting a new process, but many users aren't going to know to do this, effectively forcing them to hard-power-down the system and possibly lose open files. This is a problem from a UX perspective. There are certainly NC integrations that tie into Explorer, and this is not trivial, but there are other applications with just-as-deep integrations (e.g., Docker) that do _not_ require restarting Explorer or even restarting the computer, so it seems that this isn't a hurdle, just a "don't really care to fix".
- Even after [claims](https://github.com/nextcloud/desktop/issues/4377#issuecomment-1204818022) that reboots are no longer required, every install ends with a modal window that says _"you must restart your system for the new configurations"_.

The only progress made over the last year has been to keep it from immediately/unavoidably rebooting the machine, which while a great step-forward, is only one of many steps to improve the user experience.

I truly understand that unless we're paying for a contract with NC, the fact that we're able to use the software at all is a liberty, not a right or entitlement. _Thank you!_ I truly mean that. I feel though that many in the user-base that have meaningful bugs and UX issues are ignored so that (for instance) Hub 4 can be released tying into ChatGPT without going back to resolve system-rebooting problems from well before ChatGPT was a thing.

Please help me to understand why rebooting a system is considered necessary.

**Comment by RoyalCyan** [comment]
I had this exact bug for years. Currently I had to update to version 3.8.1 and it still required me to reboot my PC even after I consented to restarting explorer. I don't understand how is it still happening if it was supposed to be fixed in 3.5.1?

**Comment by ristein** [comment]
I just updated to v3.8.2 and clicked on "Do not close applications, a reboot will be required." and when asked to reboot I clicked on "No" (aka "I will do it later") and the app seems to be working just fine without any force close explorer or reboot.

**Comment by razorbackgeek** [comment]
> I just updated to v3.8.2 and clicked on "Do not close applications, a reboot will be required." and when asked to reboot I clicked on "No" (aka "I will do it later") and the app seems to be working just fine without any force close explorer or reboot. another issue that has been active: #4305

I really hope they've fixed it!

**Comment by r2evans** [comment]
Nope, @razorbackgeek the bug still appears on one of my laptops. I'm happy that @ristein was able to experience a reboot-free upgrade, though!

I just booted up the computer and allowed NC to update to 3.8.2. The "just booted" comment is to reflect that nothing else was running. It complained about Windows Explorer being open, so I allowed it to close Windows Explorer (and attempt to restart). Unfortunately, when the upgrade was done, Explorer is not restarted, and I have the *"You must restart your system for the configuration changes..."* modal dialog.

This has been tracked and closed in #3551, #4305, and #4377 (at least) and while (as I acknowledged before) it no longer reboots spontaneously, the need for a reboot is still unavoidable.

I wonder if there's a security policy in Win11 (perhaps Win10) that is preventing NC from being able to shift libraries/executables with a running kernel.

**Comment by ristein** [comment]
> I have the _"You must restart your system for the configuration changes..."_ modal dialog

I also had that Dialog, I just clicked No.

**Comment by r2evans** [comment]
> > I have the _"You must restart your system for the configuration changes..."_ modal dialog
> 
> I also had that Dialog, I just clicked No.

Yes, I see that to. Two other things:

1. Earlier in the installation process, it said that "open applications had to be closed", including (and today, _only_) Windows Explorer. If I choose to _not_ close the app, it says that _"a reboot will be required"_ for changes to take effect. If I choose to close the app (and restart later), _it never restarts_ leaving me without a windows manager in Windows, requiring me to go to Task Manager and manually "Run new task" to get my windows desktop functional again. Both are "poor".
2. Assuming I stopped the app (and therefore should not be required to restart), when it is done it then asks to reboot now or, if not, I will need to reboot manually for the configuration changes to take effect. There is no mention of which (if any) "configuration changes", and frankly I doubt that anything has actually been changed (full opacity of the installation/upgrade process), but it clearly insinuates that a reboot really is required even if it doesn't force you to do it.

As I said several comments earlier, the good news is that it no longer reboots spontaneously, and _that_ issue was fixed a while ago. But it still states that a reboot is required. So either (a) a reboot is required, and all of these issues complaining about reboots are still valid; (b) the reboot is not required, and nobody has bothered to update the documentation and real-time dialogs; or (c) there are some corner-cases where a reboot is actually required, and either nobody knows when they are in effect or cares to notify the rest of us.

So the problem of "requires reboot" _is still present_, at least if you believe the dialogs. There are plenty of other applications/drivers that say similar things and give you the option to reboot now or reboot later, and often those are legitimate (and credible/believable) needs to reboot. But this? There has been no acknowledgement of the core issue. If it's not required, don't say it. If it is, please make it not required or clearly explain why.

**Comment by tsa96** [comment]
I really can't emphasise enough how annoying this is, to the extend I'm considering just uninstalled the desktop software and only using the web interface.

If there's a genuine reason a reboot is required (especially given how rare it is for software to need to do this nowadays), a short sentence as to why in the update dialogue would help, along with removing the phrasing about "maybe" rebooting. I'm not going to run an update if it's going to restart my computer whilst I'm in the middle of working, nor if it's "maybe" going to restart my computer. If anything that feels even worse!

**Comment by geirawsm** [comment]
Updated to 3.9.0 just now. Got the same "restart-bug" despite allowing it to close Windows Explorer.

**Comment by budzi-2bi** [comment]
I also have the same problem, I just updated to 3.9.3 and it complained that it needs to close explorer.exe (did this for the last few automatic updates). Good thing is that you can refuse to close the apps, and then refuse to restart the computer (it all works ok from what I can see) but it is still very annoying.

**Comment by Tegenwind** [comment]
My experience is slightly different, in that if I consent it to close explorer, it does not actually reboot my computer, but won't re-open explorer.exe after finishing the installation. This leaves an end-user with an empty desktop that requires the know-how to manually start explorer.exe from there.

I've been using the options "Do not close explorer" and "restart later" for quite a while now and everything seems to work normally without rebooting, raising the question if both given options are actually necessary at all.

I've been on Windows 11 since last year, but I think my experience on Windows 10 was exactly the same before that.

**Comment by bmercernccer** [comment]
Same experience with Windows 10 just now. I don't understand the need for a reboot, as I know other shell integrations don't require one, but if this is an unresolvable limitation, can we at least provide more honest user instructions and say that a reboot will be required? Having to go through the disruptive process of closing and restarting Explorer, only to be told I have to reboot anyway makes me feel like my time has been wasted.

**Comment by munrobasher** [comment]
One suspects the installler is re-writing all the files which is what triggers a reboot. Surely it should be possible to do an update of the changes files which might just require the client to be restarted? Without delving deeply into the source code, I'm of course guessing.

**Comment by castillo92** [comment]
The problem continue in Nextcloud CLient v3.11

Setup Wizard request close Explorer.exe and then try to launch again, but this not occurr. And final step of setups, it prompt restart.

To avoid this bug, I always **not close** Explorer.exe and **not restart**. It works, but I keep on mind some day could break some funcionalities, like File Explorer integration, until a reboot occurs.

![Setup Wizard - Request close](https://github.com/nextcloud/desktop/assets/37965565/1e87c42b-b20a-4ee7-ba5d-f8d1f8f5d0d7)

**Comment by Clonkex** [comment]
Why does it need to restart Explorer at all? Other shell integrations manage to update seamlessly without restarting Explorer let alone the entire computer, so what is Nextcloud doing so wrong? Owncloud didn't have this problem when I was using that a year ago. It warned me it was about to ask for admin permissions to update, asked for admin permissions and then updated quietly in the background. No restarting of anything. I almost never update Nextcloud because it's such a pain to do.

**Comment by munrobasher** [comment]
It's **very** annoying still and the latest v3.11.0 has caused nightmare problems with a handful of upgrades. It was stuck synchronising something (can't tell as client was not responding) and worse, caused File Explorer to not respond either. It required turn off of folder redirection, uninstall (and reboot), install (and reboot), logon and wait for initial sync. The reboots just made this entire experience even more annoying. To give an idea of how annoying, the client has raised a ticket with me to look at replacing Nextcloud with something else.

**Comment by dustojnikhummer** [comment]
> Why does it need to restart Explorer at all? Other shell integrations manage to update seamlessly without restarting Explorer let alone the entire computer, so what is Nextcloud doing so wrong? Owncloud didn't have this problem when I was using that a year ago. It warned me it was about to ask for admin permissions to update, asked for admin permissions and then updated quietly in the background. No restarting of anything. I almost never update Nextcloud because it's such a pain to do.

From my experience, it doesn't. If you tell it to just do it anyway without restarting Explorer it still works.

**Comment by r2evans** [comment]
Interesting points, both from @Clonkex  and @dustojnikhummer. I _feel_ like this may no longer be an issue technologically, so if they just removed the warning and `kill_explorer.exe` function (my name), everything would continue to work unabated. They could finally close this issue and all of the other related issues (whether still-open or incorrectly-closed).

I agree with "almost never update", it's such an unnecessarily-disruptive experience that I cringed whenever the update process started for whatever reason. The user-experience should never be this disruptive.

**Comment by munrobasher** [comment]
I've been asked to go an see one of my clients who uses Nextclient. I know that replacing it is on the agenda for discussion and the disruption caused by frequent updates/restarts is one of their gripes.

**Comment by tomsik-radek** [comment]
> Interesting points, both from @Clonkex  and @dustojnikhummer. I _feel_ like this may no longer be an issue technologically, so if they just removed the warning and `kill_explorer.exe` function (my name), everything would continue to work unabated. They could finally close this issue and all of the other related issues (whether still-open or incorrectly-closed).
> 
> I agree with "almost never update", it's such an unnecessarily-disruptive experience that I cringed whenever the update process started for whatever reason. The user-experience should never be this disruptive.

I don't even mind the explorer pop-up (assuming it would be fixed so Explorer started properly), but the "you need to reboot" prompt after that just needs to die.

**Comment by r2evans** [comment]
I wonder if that's really needed? What if we don't kill explorer (i.e., the installation's "will need to restart later") and choose to not restart. Is anything not working? (Sry, no longer using windows on my personal laptop.) I wonder if this is all about aesthetics, and NC just hasn't cleaned up the #techdebt to remove the annoyance modals?

**Comment by tomsik-radek** [comment]
> @tomsik-radek , I wonder if that's really needed? What if we don't kill explorer (i.e., the installation's "will need to restart later") and choose to not restart. Is anything not working? (Sry, no longer using windows on my personal laptop.) I wonder if this is all about aesthetics, and NC just hasn't cleaned up the #techdebt to remove the annoyance modals?

I got the pop-up earlier today. Let it kill explorer (didn't start again, strike 1 for normal users) and asked me to reboot. I didn't, and it has been working for around 5 hours now. But I'm also a very light user, so maybe I just didn't encounter a potential issue.
Either way, standalone apps like nextcloud shouldn't need a reboot. If some files are held by the OS, create a second process, stop nextcloud, overwrite files, and then start it again.

**Comment by r2evans** [comment]
Okay, perhaps somebody else on this channel can try it next time?

When you update the Windows nextcloud client, try to _not_ kill `explorer.exe`, and _not_ reboot.

Once that is done, keep working as long as you can tolerate (hours, maybe a few days), and then comment back here with: (1) confirmation that you did not kill explorer and did not reboot; (2) files-sync still works; (3) explorer extensions like context-menu and VFS still works; and (4) please include your time-since-NC-update.

**Comment by ristein** [comment]
@r2evans as I said 9 months ago, this is working fine, I am doing it like this since then, no issues

**Comment by Clonkex** [comment]
I have been doing exactly that. Not letting it kill explorer, not restarting the PC, and it works fine. I very rarely restart my PCs (work or home) because I always have tonnes of things open (including explorer windows, so any attempt by NC to restart explorer is a no-go for me). That fact that I rarely restart my PCs also indicates that NC has continued working fine for a long time after updating without restarting anything.

I updated NC just now in order to test, and can absolutely confirm it's working perfectly without restarting anything.

The whole thing just feels really jank. Why do I have to tell it _twice_ not to restart explorer? Why do I have to see the installer at all? Why does it put an icon on my desktop _every single time_?

I honestly still don't get what's so special about NC. Literally no other file-syncing software is as intrusive as NC when it comes to updates. I've used OneDrive, Google Drive, Dropbox, OwnCloud and NextCloud. Out of all of them, OwnCloud and NextCloud are the only ones to require any user interaction whatsoever to update (OC asks for elevated permissions), and NextCloud is the only one that requires me to explicitly click through an installer and decide whether to restart explorer or my computer.

**Comment by munrobasher** [comment]
I don't think you'll get any disagreement from any IT support or end user on this one. Just needs fixing please Nextcloud! If there is some file that is intrinsically linked into explorer.exe like this:

explorer.exe -> nextcloud.exe

Then can't this be changed to:

explorer.exe -> nextcloudstub.exe -> nextcloud.exe

And nextcloudstub.exe never changes? Maybe I'm not getting it...

**Comment by farhills** [comment]
@claucambra @allexzander 

Apologies for the direct ping, but I think this issue needs to be acknowledged by the dev team (even if the solution is blocked for whatever reason).  I've been subscribed to this particular issue for a year, and previous issues on the topic go back to 3.4 or earlier (3+ years).  

Most of us in the discussion here are admins in some capacity, but the issue touches every single Windows user.  My site has 6 users, and it's a battle to have them update Windows and restart their machines once a week (or four).  Keeping the NC clients up to date with the current reboot headaches has been virtually impossible.

But if the answer is 'won't fix' or 'can't fix', then we'll respect the decision and move on with life.  Please give us closure! 😃 

Thanks!

**Comment by munrobasher** [comment]
Agreed... I've only got two clients with Nextcloud and I suspect that during the "special meeting" I've been asked to attend next week, we'll be taking Nextcloud out. Which is sad but one can only wait so long.

**Comment by r2evans** [comment]
@razorbackgeek, would you consider editing your initial issue post with a note to say that the consensus is currently "don't kill explorer, don't reboot, all works fine"?

@farhills and _"this particular for a year"_: for the record, there are other issues (closed, many with no resolution) for the same problem. This issue covers multiple years. https://github.com/nextcloud/desktop/issues/831, https://github.com/nextcloud/desktop/issues/3551, https://github.com/nextcloud/desktop/issues/4377

**Comment by farhills** [comment]
I did some snooping around, it looks like there may be some hope.  #6003 was assigned 3 weeks ago, and is on the dev roadmap for 2024 Spring.  I'd suggest following that issue as it seems to be the most likely solution for us right now.

**Comment by Clonkex** [comment]
@farhills Thanks for linking that, although I will forever be pissed off by people using seasons as time approximations when it depends on which half of the earth you live on (I'm Australian). They should just do Q1, Q2 etc. instead.

**Comment by RegisPerdreau** [comment]
Hi,
I think the bug or lack of functionality is still pending. With version 3.13, windows 11, the Nextcloud installer asks to close  Explorer and Nextcloud Desktop and can't close both even if the previous version is 3.12.5. So the last choice is to reboot as usual.

**Comment by schwimmcoder** [comment]
Bug still exists while updating from 3.13.2 to 3.13.4

**Comment by willemb2** [comment]
Still present in 3.14.1 and extremely confusing and annoying to our end users. To me it seems unnecessary, because you see the screen turn black and taskbar disappear, which I believe means Windows Explorer was closed successfully.

**Comment by enrikre** [comment]
Bug ist still persisting, generally it is annoying to have a update feature that requires 15 clicks every few weeks, and additionally restarts, and, as I read here, even crashes the system. These sort of bugs make me consider to move away from NC. See also the forum threats about quite a large number of unhappy (ex)users.

**Comment by munrobasher** [comment]
> Also would like for the windows client update process to be simpler (ideally just the 'proceed with admin priv|give me admin priv' ok? click).

I was rather pleased to see that the venerable Notepad++ made their update mechanism even simpler. It wasn't hard before now but they've added a "Silent" option to speed up dates.

**Comment by willemb2** [comment]
In addition to Explorer it sometimes complains about "COM Surrogate". Yes, I know what that is, but the name alone is enough to scare the hell out my end users. "Is Nextcloud spreading a virus?"
This annoying client update bug is still present in 3.15.3 but in general Nextcloud is improving step by step and I noticed it recovers from Windows sleep much quicker since some recent update.

**Comment by chinggg** [comment]
I can confirm this problem still exists on my Windows machine. Worse still, Nextcloud does not even update to the latest version. After updating & rebooting once, it immediately reminds me that there is a new version available, and I have to reboot again 😰

**Comment by razorbackgeek** [comment]
I found a workaround for this annoyance.  If you use winget to update, it doesn't force you to reboot.

**Comment by jquiros2** [comment]
Noice.  Confirming this works better than the default.
winget upgrade --id=Nextcloud.NextcloudDesktop -e

**Comment by reox** [comment]
> ` winget upgrade --id=Nextcloud.NextcloudDesktop -e`

that is a nice idea and I can confirm that there is no reboot dialog. However, the nextcloud client did not start automatically after the update - thus don't forget to start it manually afterwards.

**Comment by nilsding** [comment]
the Winget mention was a really good hint, thanks @razorbackgeek!

took a look at [the manifest](https://github.com/microsoft/winget-pkgs/blob/0b40904d1480f67d48c6d8e2ac7f27b31aa566c7/manifests/n/Nextcloud/NextcloudDesktop/3.16.5/Nextcloud.NextcloudDesktop.installer.yaml#L16C11-L16C32) and saw that the [`REBOOT` MSI property](https://learn.microsoft.com/en-us/windows/win32/msi/reboot) is set to `ReallySuppress`.  I'll try to set that in the MSI directly, let's see if that'll work...

**Comment by nilsding** [comment]
the installer for 3.16.6 and 3.17.0 (and future daily builds `>= 20250605`) will have this MSI property set by default

I'll leave this issue open until we collected enough feedback of the restart prompt no longer appearing

**Comment by addy90** [comment]
I had for the first time ever no reboot necessary when upgrading Nextcloud client.
It was this time the automatic upgrade from 3.16.5 to 3.16.6.

I was so amazed and baffled that I googled to find this issue and leave a comment that in my case, it worked!

Thank you folks!

**Comment by willemb2** [comment]
Yes, it looks like it has been fixed in [3.16.6](https://github.com/nextcloud/desktop/releases/tag/v3.16.6) Hurray 👏
Off-topic: I'm not complaining because I use the free open source version of NC, but I sometimes wonder what effect long lasting bugs and annoyances like this have on enterprise customers.

**Comment by schwimmcoder** [comment]
Just updated to 3.16.6 and no Restart was necessary, awesome! 

Thanks for everyone to fix this!

**Comment by castillo92** [comment]
I just update to v3.16.6 and checked issue is solved. Thanks!!

**Comment by mgallien** [comment]
this has been fixed already
